features/environment.py

```diff
@@ -1,294 +1,340 @@
+                return f.read().strip()
+        except IOError:
+            return None
+
+    def _start(self):
+        return subprocess.Popen(['coverage', 'run', '--source=patroni', '-p', 'patroni.py', self._config],
+                                stdout=self._log, stderr=subprocess.STDOUT, cwd=self._work_directory)
 
-    def _is_running(self, pg_name):
-        return pg_name in self._processes and self._processes[pg_name].pid and (self._processes[pg_name].poll() is None)
+    def _is_accessible(self):
+        return self.query("SELECT 1", fail_ok=True) is not None
 
-    def _make_patroni_test_config(self, pg_name, tags=None):
-        patroni_config_name = PatroniController.PATRONI_CONFIG.format(pg_name)
+    def _make_patroni_test_config(self, name, dcs, tags):
+        patroni_config_name = self.PATRONI_CONFIG.format(name)
         patroni_config_path = os.path.join(self._output_dir, patroni_config_name)
 
         with open(patroni_config_name) as f:
             config = yaml.load(f)
-        postgresql = config['postgresql']
-        postgresql['name'] = pg_name
-        postgresql['data_dir'] = 'data/{0}'.format(pg_name)
-        postgresql_params = postgresql['parameters']
-        postgresql_params['logging_collector'] = 'on'
-        postgresql_params['log_destination'] = 'csvlog'
-        postgresql_params['log_directory'] = self._output_dir
-        postgresql_params['log_filename'] = '{0}.log'.format(pg_name)
-        postgresql_params['log_statement'] = 'all'
-        postgresql_params['log_min_messages'] = 'debug1'
-        postgresql_params['unix_socket_directories'] = '.'
+
+        self._connstring = self._make_connstring(config)
+
+        config['postgresql'].update({'name': name, 'data_dir': self._data_dir})
+        config['postgresql']['parameters'].update({
+            'logging_collector': 'on', 'log_destination': 'csvlog', 'log_directory': self._output_dir,
+            'log_filename': name + '.log', 'log_statement': 'all', 'log_min_messages': 'debug1'})
 
         if tags:
             config['tags'] = tags
 
+        if dcs != 'etcd':
+            etcd = config.pop('etcd')
+            config['zookeeper'] = {'scope': etcd['scope'], 'session_timeout': etcd['ttl'],
+                                   'reconnect_timeout': config['loop_wait']}
+            if dcs == 'exhibitor':
+                config['zookeeper']['exhibitor'] = {'hosts': ['127.0.0.1'], 'port': 8181}
+            else:
+                config['zookeeper']['hosts'] = ['127.0.0.1:2181']
+
         with open(patroni_config_path, 'w') as f:
             yaml.dump(config, f, default_flow_style=False)
 
         return patroni_config_path
 
-    def _make_connstring(self, pg_name):
-        if pg_name in self._connstring:
-            return self._connstring[pg_name]
+    @staticmethod
+    def _make_connstring(config):
+        tmp = (config['postgresql']['connect_address'] + ':5432').split(':')
```

Comment:
Should be [:2], otherwise, tmp might have 3 members while the string below expects only 2.